lib/go/thrift/simple_server.go

```diff
@@ -193,7 +195,13 @@ func (p *TSimpleServer) innerAccept() (int32, error) {
 	}
 	if client != nil {
 		p.wg.Add(1)
+		p.clients[client] = struct{}{}
 		go func() {
+			defer func() {
+				p.mu.Lock()
+				defer p.mu.Unlock()
+				delete(p.clients, client)
```

Comment:
Go has a runtime bug that a builtin map never actually shrinks when you delete from it: https://github.com/golang/go/issues/20135

Which means although the behavior is correct (you won't try to close those clients that's already disconnected during `Close`), for a long running server this map will keep growing indefinitely and eventually causing memory issues.

For that reason, a `sync.Map` will probably actually be better, as I understood it `sync.Map` actually occasionally replaces the whole map under the hood so it actually can be garbage collected.

Also I'm not sure that you can actually use a `TTransport` as the map key (for either builtin map or `sync.Map`), when you try to use a type that cannot be used as the map key it will cause a runtime panic instead of some compile time error. We might need to make some breaking change to expose the `TSocket`/`TSSLSocket` that's wrapped inside this `TTransport` and use its `.Addr().String()` as the map key (and ignore any `TTransport` that does not have a `TSocket`/`TSSLSocket` under the hood).

Reply:
> `sync.Map` will probably actually be better, as I understood it `sync.Map` actually occasionally replaces the whole map under the hood so it actually can be garbage collected.

Looking at the implementation, I don't think `sync.Map` would help either. The underlying "dirty" map is only replaced after a number of `Load()` misses, which would never happen here as we only read once during `Close()`.

> Also I'm not sure that you can actually use a TTransport as the map key (for either builtin map or sync.Map), when you try to use a type that cannot be used as the map key it will cause a runtime panic instead of some compile time error.

I'm not seeing any Go test failures on Travis (though I didn't check thoroughly) so I assume all implementations of `TTransport` are hashable.

Reply:
Looks like at least a `*TSocket` is comparable:

```
$ cat main.go 
package main

import (
        "fmt"

        "github.com/apache/thrift/lib/go/thrift"
)

func main() {
        sock := thrift.NewTSocketConf("localhost:6060", nil)
        m := map[thrift.TTransport]struct{}{
                sock: {},
        }
        fmt.Println(m)
}
$ go run main.go 
map[0xc000068120:{}]
```

But still, there's no guarantee that `TTransport` returned by `TServerSocket.Accept()` will always be comparable, and when it's not trying to use it as a key will cause a runtime panic and it will be hard to detect in tests. (For example, we have a pending PR to wrap `TServerSocket` to return a `TTransport` that counts the number of active clients: https://github.com/reddit/baseplate.go/pull/464)

So if this is really important to you, make it opt-in instead: This map is only used when people call an exported function from `TSimpleServer` to enable it (for example, `TSimpleServer.EnableClientCleanup()`, name tbd as I'm bad at names). And we should note in the function's doc about the implications:

1. The memory used by this map could be unbounded
2. If you enable this map, `TTransport` returned by your `TServerTransport` must be comparable or panics will happen

Reply:
Even with the opt-in approach, consider how big a footgun using `TTransport` as the map key could be, I think it's still better to:

1. Make an breaking API change, add `Addr() net.Addr` to `TTransport` interface (`TMemoryBuffer.Addr()` should return `nil`, delegate transports like `TFramedTransport`, `TBufferedTransport`, `THeaderTransport`, etc. should just call their wrapped `TTransport.Addr()`)
2. We use `TTransport.Addr().String()` as the map key, and don't put the `TTransport` into the map if `Addr()` returns `nil` (this should only happen when we use `TMemoryBuffer` under the hood, I think, which should be fine)

@dcelasun what do you think?

Another alternative is to add something like `TTransport.String()` to the interface, but the behavior of that would be more undefined.

Reply:
Another idea I had considered before for a different issue is to add an `Unwrap() TTransport` api to `TTransport`, similar to how `errors.Unwrap()` works. This way we can keep unwrapping until we find the `*TSocket` or `*TSSLSocket` wrapped under the hood and use its `Addr().String()` as the map key (and if there's no `*TSocket`/`*TSSLSocket`, then we don't put this client into the map).

Reply:
Breaking `TTransport` will definitely be painful. When we added `context.Context` to function signatures, it broke so many downstream projects. This won't be as bad, but still.

Anyway, as long as we're adding new API to `TTransport` I like `Addr() net.Addr` better than `Unwrap() TTransport`. It feels more readable.

Reply:
I'm also not a fan of the opt-in or documentation approaches. The former, because the current behaviour is wrong and should be fixed, and the latter because I don't think end users should worry about an OOM caused by an internal implementation detail.

Reply:
I think depending on the real world situation "memory of a server grows unbounded and will eventually OOM" could be more severe than "the shutdown of the server will take forever and eventually timeout". But this is a fair point:

>I'm also not a fan of the opt-in or documentation approaches. The former, because the current behaviour is wrong and should be fixed, and the latter because I don't think end users should worry about an OOM caused by an internal implementation detail.

But in order to turn it default on, we must make the memory bounded. One way to do it is to add a background goroutine with `TSimpleServer` that use a ticker to copy the map over regularly, but when doing so it also need to hold the lock, and preventing `TSimpleServer` from processing any new requests (so it's kind of like a stop-the-world gc). The goroutine function should be something like this:

```go
func (p *TSimpleServer) mapCleanup() {
  p.mu.Lock()
  defer p.mu.Unlock()
  newMap := make(map[string]TTransport, len(p.clients))  // the actual type of the map pending on which type we decide in the end
  for k, v := range p.clients {
    newMap[k] = v
  }
  p.clients = newMap
}
```

It might not be too bad? This copy should be reasonably efficient in go, but I would like to see some real world impact first before merging it.

Reply:
Or maybe instead of a ticker, make it run every N deletes. I'm not sure which would actually work better in real world scenarios.

Reply:
> Or maybe instead of a ticker, make it run every N deletes. I'm not sure which would actually work better in real world scenarios.

I think every N deletes is better because with a ticker it can still OOM during traffic spikes.

Reply:
with either a ticker or every N deletes, we should make the ticker time/N configurable.

so if we choose ticker approach, and some service know that it could handle a lot of connections, they can choose to set a shorter ticker time to cleanup more often.

I think a shorter ticker time might work better as with a shorter ticker time usually also comes with smaller map to copy over, so the whole "stop-the-world-gc-ish" thing can be quicker and less noticeable.

Reply:
now I wonder how stdlib http server resolve this issue. maybe I should take a look at its implementation.

Reply:
well, for http server (https://pkg.go.dev/net/http#Server.Shutdown):

>When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return.
>Shutdown does not attempt to close nor wait for hijacked connections such as WebSockets. The caller of Shutdown should separately notify such long-lived connections of shutdown and wait for them to close, if desired. See RegisterOnShutdown for a way to register shutdown notification functions.

Reply:
instead of either of the stop-the-world-gc-ish approaches, a better approach might be to make the read not blocking forever. basically we could try to always add a timeout to the read, and when timeout happens check if the server is closed, and if not, we can go back to read again.

this might need some compiler change to accomplish, though. as it might need some changes to the processor. haven't really looked into the details yet.

Reply:
Actually, I misinterpreted that go runtime bug. It's just that the capacity of the map only grows and never shrinks (so delete won't cause the map to shrink), but memory holding deleted items can be reused for future inserts. So although this is still unbounded, it will be more like a high watermark of how many concurrent connection the server handles, and unlikely to cause OOM. So using a map is fine. But we still need to handle the map key issue.

This is the code I used to verify the behavior (note the different allocations between the 2 benchmark tests):

```
$ go test -v -bench .
=== RUN   TestHelpers
    main_test.go:43: size: 1000
    main_test.go:46: size: 502
--- PASS: TestHelpers (0.00s)
goos: linux
goarch: amd64
pkg: test
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkMap
BenchmarkMap/fill-twice
    main_test.go:60: total size 2000, n 1, average size: 2000
    main_test.go:60: total size 200000, n 100, average size: 2000
    main_test.go:60: total size 5982000, n 2991, average size: 2000
    main_test.go:60: total size 14522000, n 7261, average size: 2000
BenchmarkMap/fill-twice-12                  7261            184987 ns/op          101042 B/op        113 allocs/op
BenchmarkMap/fill-delete-fill
    main_test.go:72: total size 1505, n 1, average size: 1505
    main_test.go:72: total size 150243, n 100, average size: 1502
    main_test.go:72: total size 8163265, n 5442, average size: 1500
    main_test.go:72: total size 12801894, n 8534, average size: 1500
BenchmarkMap/fill-delete-fill-12            8534            146318 ns/op           50348 B/op         91 allocs/op
PASS
ok      test    3.927s
$ cat main_test.go 
package main

import (
        "math/rand"
        "os"
        "testing"
)

const (
        n = 1000

        margin = 0.1
)

type mapType = map[int]struct{}

func fillMap(m mapType, n int) {
        for i := 0; i < n; i++ {
                m[rand.Int()] = struct{}{}
        }
}

func deleteHalf(m mapType) {
        for k := range m {
                if rand.Float64() >= 0.5 {
                        delete(m, k)
                }
        }
}

func compareMargin(tb testing.TB, want, got int, margin float64) {
        tb.Helper()

        diff := int(float64(want) * margin)
        if got < want-diff || got > want+diff {
                tb.Errorf("Want %d got %d margin %v", want, got, margin)
        }
}

func TestHelpers(t *testing.T) {
        m := make(mapType)
        fillMap(m, n)
        t.Logf("size: %d", len(m))
        compareMargin(t, n, len(m), margin)
        deleteHalf(m)
        t.Logf("size: %d", len(m))
        compareMargin(t, n/2, len(m), margin)
}

func BenchmarkMap(b *testing.B) {
        b.Run("fill-twice", func(b *testing.B) {
                b.ReportAllocs()
                var totalSize int
                for i := 0; i < b.N; i++ {
                        m := make(mapType)
                        fillMap(m, n)
                        fillMap(m, n)
                        totalSize += len(m)
                }
                b.Logf("total size %d, n %d, average size: %v", totalSize, b.N, totalSize/b.N)
        })
        b.Run("fill-delete-fill", func(b *testing.B) {
                b.ReportAllocs()
                var totalSize int
                for i := 0; i < b.N; i++ {
                        m := make(mapType)
                        fillMap(m, n)
                        deleteHalf(m)
                        fillMap(m, n)
                        totalSize += len(m)
                }
                b.Logf("total size %d, n %d, average size: %v", totalSize, b.N, totalSize/b.N)
        })
}

func TestMain(m *testing.M) {
        rand.Seed(1)
        os.Exit(m.Run())
}
```

Reply:
another potential issue with the approach: closing the socket in one goroutine while another goroutine is reading from/writing to the socket might cause race conditions?

Reply:
> another potential issue with the approach: closing the socket in one goroutine while another goroutine is reading from/writing to the socket might cause race conditions?

from the doc of `net.Conn`:
```
	// Close closes the connection.
	// Any blocked Read or Write operations will be unblocked and return errors.
	Close() error
```

so I guess this is fine.

But in that case we don't even need a map? we can just add one additional goroutine per client connection, basically:

1. In `TSimpleServer.Serve` we create a `context.WithCancel` and store both the ctx and cancel function `TSimpleServer`
2. cancel that ctx in `TSimpleServer.Close`
3. in every client accept goroutine, we start another goroutine to be unblocked by the server ctx to close the connection, basically changing https://github.com/apache/thrift/blob/999e6e3bce217acb35b44440fd656cf169d47ed8/lib/go/thrift/simple_server.go#L196-L201 into:

```go
		go func() {
			defer p.wg.Done()
			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()
			go func() {
				select {
				case <-ctx.Done():
					// client exited, do nothing
				case <- p.ctx.Done():
					// TSimpleServer.Close called, close the client connection
					client.Close()
				}
			}()
			if err := p.processRequests(client); err != nil {
				p.logger(fmt.Sprintf("error processing request: %v", err))
			}
		}()
```

basically we trade a (kinda unbounded) map with one additional goroutine per client connection.

Reply:
> we trade a (kinda unbounded) map with one additional goroutine per client connection. the big advantage is that we no longer need to make a breaking change to TTransport to make it usable in a map.

Sounds much more reasonable. @buptubuntu what do you think?

Reply:
I have also considered to add one additional goroutine per client connection，but not sure if that will cause performance problem. But if we want to fix the map key issue we will have to make big change to TTrancsport， that will be another issue that we will handle in the future.Besides the cost of the additinal go routine should be quite small comparing to the cost of user defined processor or the data read and write operations in transport.

So I think this is a better solution theoretically

I will try to get the cost of the additional goroutine for some extreme cases:
1>1000~10000 client connect to server but not send data for a long time
2>1000~10000 client connect to server and send 1k data per second

Reply:
I have tested the mem the goroutine cost using the code below:
```golang
package main

import (
	"context"
	"fmt"

	"runtime"
	"sync"
)

func main() {
	memConsumed := func() uint64 {
		runtime.GC()
		var s runtime.MemStats
		runtime.ReadMemStats(&s)
		return s.Sys
	}

	var channel <-chan interface{}
	var wg sync.WaitGroup

	const numGoroutines = 1000000 // 1M
	wg.Add(numGoroutines)

	before := memConsumed()
	for i := numGoroutines; i > 0; i-- {
		ctx, cancel := context.WithCancel(context.Background())
		go func() {
			wg.Done()
			cancel()
			<-ctx.Done()
			<-channel

		}()
	}
	wg.Wait()
	after := memConsumed()

	fmt.Printf("%.3fkb", float64(after-before)/numGoroutines/1024)
}
```

cost as below:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/buptubuntu/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/buptubuntu/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
</head>

<body link="#0563C1" vlink="#954F72">



goroutines | total mem | mem per goroutine
-- | -- | --
10000 | 35M | 3.5K
1000 | 5.3M | 5.3K



</body>

</html>

Reply:
I have also tested the cost per client using code below:

```golang
package main

import (
	"context"
	"fmt"
	"net"
	_ "net/http/pprof"
	"runtime"
	"strings"
	"time"

	"github.com/apache/thrift/lib/go/thrift"
)

func main() {
	memConsumed := func() uint64 {
		runtime.GC()
		var s runtime.MemStats
		runtime.ReadMemStats(&s)
		return s.Sys
	}

	ln, _ := net.Listen("tcp", "localhost:0")

	proc := &mockProcessor{
		ProcessFunc: func(in, out thrift.TProtocol) (bool, thrift.TException) {
			in.ReadMessageBegin(context.Background())
			time.Sleep(time.Hour)
			return false, nil
		},
	}

	trans := &mockServerTransport{
		ListenFunc: func() error {
			return nil
		},
		AcceptFunc: func() (thrift.TTransport, error) {
			conn, err := ln.Accept()
			if err != nil {
				// t.Errorf("error accept connection")
				return nil, err
			}
			return thrift.NewTSocketFromConnTimeout(conn, 0), nil
		},
		CloseFunc: func() error {
			return nil
		},
		InterruptFunc: func() error {
			return ln.Close()
		},
	}

	serv := thrift.NewTSimpleServer2(proc, trans)
	go serv.Serve()
	time.Sleep(time.Second)
	const numGoroutines = 1000
	conns := make([]net.Conn, 0, numGoroutines)
	before := memConsumed()

	for i := 0; i < numGoroutines; i++ {
		port := strings.Split(ln.Addr().String(), ":")[1]
		netConn, _ := net.Dial("tcp", "localhost:"+port)
		time.Sleep(time.Millisecond)
		conns = append(conns, netConn)
	}

	after := memConsumed()

	time.Sleep(time.Second)
	fmt.Printf("%.3fkb", float64(after-before)/numGoroutines/1024)
}

type mockProcessor struct {
	ProcessFunc func(in, out thrift.TProtocol) (bool, thrift.TException)
}

func (m *mockProcessor) Process(ctx context.Context, in, out thrift.TProtocol) (bool, thrift.TException) {
	return m.ProcessFunc(in, out)
}

func (m *mockProcessor) ProcessorMap() map[string]thrift.TProcessorFunction {
	return map[string]thrift.TProcessorFunction{
		"mock": thrift.WrappedTProcessorFunction{
			Wrapped: func(ctx context.Context, seqId int32, in, out thrift.TProtocol) (bool, thrift.TException) {
				return m.ProcessFunc(in, out)
			},
		},
	}
}

func (m *mockProcessor) AddToProcessorMap(name string, processorFunc thrift.TProcessorFunction) {}

type mockWrappedProcessorContextKey int

const (
	processorName mockWrappedProcessorContextKey = iota
)

// setMockWrappableProcessorName sets the "name" of the TProcessorFunction to
// call on a mockWrappableProcessor when calling Process.
//
// In a normal TProcessor, the request name is read from the request itself
// which happens in TProcessor.Process, so it is not passed into the call to
// Process itself, to get around this in testing, mockWrappableProcessor calls
// getMockWrappableProcessorName  to get the name to use from the context
// object.
func setMockWrappableProcessorName(ctx context.Context, name string) context.Context {
	return context.WithValue(ctx, processorName, name)
}

// getMockWrappableProcessorName gets the "name" of the TProcessorFunction to
// call on a mockWrappableProcessor when calling Process.
func getMockWrappableProcessorName(ctx context.Context) (string, bool) {
	val, ok := ctx.Value(processorName).(string)
	return val, ok
}

// mockWrappableProcessor can be used to create a mock object that fufills the
// TProcessor interface in testing.
type mockWrappableProcessor struct {
	ProcessorFuncs map[string]thrift.TProcessorFunction
}

// Process calls the TProcessorFunction assigned to the "name" set on the
// context object by setMockWrappableProcessorName.
//
// If no name is set on the context or there is no TProcessorFunction mapped to
// that name, the call will panic.
func (p *mockWrappableProcessor) Process(ctx context.Context, in, out thrift.TProtocol) (bool, thrift.TException) {
	name, ok := getMockWrappableProcessorName(ctx)
	if !ok {
		panic("MockWrappableProcessorName not set on context")
	}
	processor, ok := p.ProcessorMap()[name]
	if !ok {
		panic(fmt.Sprintf("No processor set for name %q", name))
	}
	return processor.Process(ctx, 0, in, out)
}

func (p *mockWrappableProcessor) ProcessorMap() map[string]thrift.TProcessorFunction {
	return p.ProcessorFuncs
}

func (p *mockWrappableProcessor) AddToProcessorMap(name string, processorFunc thrift.TProcessorFunction) {
	p.ProcessorFuncs[name] = processorFunc
}

var (
	_ thrift.TProcessor = (*mockProcessor)(nil)
	_ thrift.TProcessor = (*mockWrappableProcessor)(nil)
)

type mockServerTransport struct {
	ListenFunc    func() error
	AcceptFunc    func() (thrift.TTransport, error)
	CloseFunc     func() error
	InterruptFunc func() error
}

func (m *mockServerTransport) Listen() error {
	return m.ListenFunc()
}

func (m *mockServerTransport) Accept() (thrift.TTransport, error) {
	return m.AcceptFunc()
}

func (m *mockServerTransport) Close() error {
	return m.CloseFunc()
}

func (m *mockServerTransport) Interrupt() error {
	return m.InterruptFunc()
}

type mockTTransport struct {
	thrift.TTransport
}

func (m *mockTTransport) Close() error {
	return nil
}
```
cost as below:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/buptubuntu/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/buptubuntu/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
</head>

<body link="#0563C1" vlink="#954F72">



clients | total mem | mem per client
-- | -- | --
10000 | 157M | 15.7K
1000 | 24M | 24K



</body>

</html>

Reply:
For the cpu cost, as the operation of the additional goroutine is quite simple(close the client) and we just move it from the stop function,i think we can just ignore that